Added support for negative axis in GpuCumsumOp

MarcCote opened the pull request:
When a negative axis is provided to `GpuCumsumOp`, `CumsumOp` or `CumprodOp`, it will be converted to its positive equivalent (inferred from the given input array). 

Since the axis conversion happens during ~~the making of the node~~ the generation of the C code, two `cumsum` nodes created using a 2D array and respectively `axis=1` and `axis=-1` ~~will~~ won't be equivalent (their hash ~~will~~ won't be the same).~~

Closes #2715 

Edit: Saying two `cumsum` nodes with different axis may be equivalent was based on a wrong assumption.

abergeron commented on theano/sandbox/cuda/extra_ops.py:
```diff
@@ -37,8 +37,11 @@ def make_node(self, x):
             raise ValueError('axis(={1}) out of bounds'.format(self.axis))
+        elif self.axis < 0:
+            # Convert negative axis to positive axis.
+            self.axis += x.ndim
```
This is wrong.  'axis' is global for all places that use this op instance, not just for the node you are building right now.

MarcCote replied:
I was afraid of that. I'll move it in c_code then.

abergeron commented on theano/sandbox/cuda/extra_ops.py:
```diff
@@ -352,7 +355,10 @@ def c_support_code_apply(self, node, nodename):
+
+        # We assume array has been already flattened if needed.
         axis = self.axis if self.axis is not None else 0
+
```
You should do your modifications of axis here instead.

MarcCote replied:
Is there a way I can infer the number of dimension of from x? If not, I'll modify the string `code` to use `CudaNdarray_NDIM(%(x)s)`

MarcCote commented:
@abergeron I moved the conversion of negative axis to positive axis inside the method `c_code`. More precisely, I moved it inside the C code. I also removed the conversion from `CumsumOp` and `CumprodOp` since they support negative axis natively.

abergeron commented:
This looks good now.
